sympy/functions/elementary/trigonometric.py

```diff
@@ -107,20 +107,25 @@ def _peeloff_pi(arg):
     (x + pi*y + pi/6, pi/2)
 
     """
+    pi_coeff = S.Zero
+    rest_terms = []
     for a in Add.make_args(arg):
-        if a is S.Pi:
-            K = S.One
-            break
-        elif a.is_Mul:
-            K, p = a.as_two_terms()
-            if p is S.Pi and K.is_Rational:
-                break
-    else:
+        K = a.coeff(S.Pi)
+        if K and K.is_rational:
+            pi_coeff += K
+        else:
+            rest_terms.append(a)
+
+    if pi_coeff is S.Zero:
         return arg, S.Zero
 
-    m1 = (K % S.Half)*S.Pi
-    m2 = K*S.Pi - m1
-    return arg - m2, m2
+    m1 = (pi_coeff % S.Half)*S.Pi
+    m2 = pi_coeff*S.Pi - m1
+    final_coeff = m2 / S.Pi
+    if final_coeff.is_rational:
+        if not (sin(m2).has(sin) and cos(m2).has(cos)):
+            return Add(*(rest_terms + [m1])), m2
```

Comment:
Since this function is called by e.g. `sin(expr)` calling `sin` or `cos` again risks infinite recursion.

Why are these calls needed? Presumably there is some logic in `sin.eval` that helps out here but it would probably be better to use that logic directly rather than call into `sin` and `cos` again.